### PR TITLE
[Sketch]Module for digital signal processing

### PR DESCRIPTION
**Pull request by SusanDoggie**
issue #46 

Considerations:

1. Name of module that will not conflict to Apple’s library.
2. What are the rules for naming the functions?
3. What functions need?

Do we need these:

- Elementwise Vector Arithmetic
  - add, sub, mul, div
- Vector-Scalar Arithmetic
- Vector-Vector Arithmetic
- Vector Reduction
- Vector Geometry Functions
- 2D FFT and convolution

TODO:

- [x] radix 2 cooley-tukey
- [x] fast convolution

### CODE REVIEW

**Comment by SusanDoggie** [comment]
> Why the path `Source/Accelerate/Implements`? Maybe "Implementation", but neither target `Complex` nor `Real` uses this file structure.
> 

The functions requires the Real module for sin, cos functions. I am not sure should we implement the wrapper function for the non-splitting complex buffers.

for example:

```
public func vDSP_fft_zop(_ log2N: Int, _ input: UnsafePointer<Complex>, _ in_stride: Int, _ in_count: Int, _ output: UnsafeMutablePointer<Complex>, _ out_stride: Int) {
    input._reboundToDouble { _input in
 output._reboundToDouble { _output  in
 vDSP_fft_zop(log2N, _input, _input.successor(), in_stride << 1, in_count, _output, _output.successor(), out_stride << 1)
 }
 }
}
```

File: Tests/AccelerateTests/FastFourierTransformTests.swift [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    
+    return result
+}
+
+final class FastFourierTransformTests: XCTestCase {
+    
+    let accuracy = 0.00000001
+    
+    func test_vDSP_fft_zrop_2() {
```
[the remark this answers is not in the document]
Comment (SusanDoggie) [reply Tests/AccelerateTests/FastFourierTransformTests.swift]: updated: https://github.com/apple/swift-numerics/pull/44/commits/c0fe2189ead4f430ee828bc4e7073840ac731bb6

File: Sources/Accelerate/Implements/vDSP_fft_zop_imp_2.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Real
+
+@inlinable
+@inline(__always)
+func vDSP_fft_zop_imp_2<T: Real>(_ real: UnsafePointer<T>, _ imag: UnsafePointer<T>, _ in_stride: Int, _ _real: UnsafeMutablePointer<T>, _ _imag: UnsafeMutablePointer<T>, _ out_stride: Int) {
```
[the remark this answers is not in the document]
Comment (SusanDoggie) [reply Sources/Accelerate/Implements/vDSP_fft_zop_imp_2.swift]: should i define a typealias for that?

**Comment by SusanDoggie** [comment]
All the function completed
Should we rename all the function for clarity?

Here is the function list
vDSP_fft_zrip    // Real to Complex In-place FFT
vDSP_fft_zrop    // Real to Complex Out-of-place FFT
vDSP_fft_zip    // Complex to Complex In-place FFT
vDSP_fft_zop    // Complex to Complex Out-of-place FFT

**Comment by stephentyrone** [comment]
Hi Susan!

First, I'm super-excited to see this. Structurally, there are two issues to address with this PR before we get any further:

- This would be a new module, so it should be pitched as an issue, and then developed on a branch so that people can experiment for a while, rather than merged straight to master.
- Accelerate is an Apple SDK framework, so we wouldn't use that name in this repo (for both copyright reasons and to avoid confusion).

What I'd like you to do: raise an issue to pitch adding an "FFT" (or maybe "Transforms"--would we also have a DCT etc. eventually?) module. You don't need to put a ton of work into this right away, just a couple sentences explaining what you want to add and why. Naming and scope can be discussed on that thread.

Once that discussion happens, I'll create a branch that work can happen on. Some quick other notes, based only on skimming the source:

- Please avoid copying the vDSP naming. Both to avoid confusion with the Accelerate implementation (which is also available in Swift on Apple platforms), and because they are objectively bad names, inherited from a time when characters in identifiers were precious commodities hoarded as generational wealth.

- These vDSP-style interfaces simply don't import very well into Swift. You've improved them a bit, but a module should really have public interfaces that avoid `UnsafeBufferPointer` as much as possible without compromising performance.

Again, I'm excited to see work in this direction, so open an issue to discuss adding a module, and we'll go from there.

**Comment by SusanDoggie** [comment]
> Hi Susan!
> 
> First, I'm super-excited to see this. Structurally, there are two issues to address with this PR before we get any further:
> 
> * This would be a new module, so it should be pitched as an issue, and then developed on a branch so that people can experiment for a while, rather than merged straight to master.
> * Accelerate is an Apple SDK framework, so we wouldn't use that name in this repo (for both copyright reasons and to avoid confusion).
> 
> What I'd like you to do: raise an issue to pitch adding an "FFT" (or maybe "Transforms"--would we also have a DCT etc. eventually?) module. You don't need to put a ton of work into this right away, just a couple sentences explaining what you want to add and why. Naming and scope can be discussed on that thread.
> 
> Once that discussion happens, I'll create a branch that work can happen on. Some quick other notes, based only on skimming the source:
> 
> * Please avoid copying the vDSP naming. Both to avoid confusion with the Accelerate implementation (which is also available in Swift on Apple platforms), and because they are objectively bad names, inherited from a time when characters in identifiers were precious commodities hoarded as generational wealth.
> * These vDSP-style interfaces simply don't import very well into Swift. You've improved them a bit, but a module should really have public interfaces that avoid `UnsafeBufferPointer` as much as possible without compromising performance.
> 
> Again, I'm excited to see work in this direction, so open an issue to discuss adding a module, and we'll go from there.

Yes, I agree that we should rename all the functions. It's a draft for now.

**Comment by CTMacUser** [comment]
How big of a deal is it for the FFT functions to work on memory buffers, instead of generic `(Mutable)Collection`?  If the algorithms are workable with general collections, maybe we could just do that and drop the accelerate-buffer protocols.

If we switch to general collections, we could use sub-sequences to describe portions of a collection, move the power-of-2 check to see if a sub-sequence’s `count` has one set bit, and transition the stride parameter to using collection wrappers that automatically skip elements.  But some of these changes are likely not worth it, especially for the effort required.

**Comment by SusanDoggie** [comment]
> How big of a deal is it for the FFT functions to work on memory buffers, instead of generic `(Mutable)Collection`? If the algorithms are workable with general collections, maybe we could just do that and drop the accelerate-buffer protocols.
> 
> If we switch to general collections, we could use sub-sequences to describe portions of a collection, move the power-of-2 check to see if a sub-sequence’s `count` has one set bit, and transition the stride parameter to using collection wrappers that automatically skip elements. But some of these changes are likely not worth it, especially for the effort required.

Here are my working examples
https://github.com/SusanDoggie/Doggie/blob/master/Sources/Doggie/ImageFilter/ImageConvolution.swift

If we given a raw pixels in this format
[red, green , blue, alpha, red, green , blue, alpha, red, green , blue, alpha, ...]

It is easy to separate the interlaced channels when you using raw pointer with stride parameter without copy
red: [red, _ , _, _, red, _ , _, _, red, _ , _, _, ...]
green: [_, green , _, _, _, green , _, _, _, green , _, _, ...]
blue: [_, _ , blue, _, _, _ , blue, _, _, _ , blue, _, ...]
alpha: [_, _ , _, alpha, _, _ , _, alpha, _, _ , _, alpha, ...]

**Comment by stephentyrone** [comment]
> It is easy to separate the interlaced channels when you using raw pointer and stride parameter without copy
> red: [ red, _ , _, _, red, _ , _, _, red, _ , _, _, ... ]
> green: [ _, green , _, _, _, green , _, _, _, green , _, _, ... ]
> blue: [ _, _ , blue, _, _, _ , blue, _, _, _ , blue, _, ...]
> alpha: [ _, _ , _, alpha, _, _ , _, alpha, _, _ , _, alpha, ... ]

This particular thing wouldn't stop you from using Collections; you could pretty easily vend a strided view of a base collection like this (just like a slice). But we don't actually *want* to encourage operating on strided buffers like this in general; it's a performance hazard if you provide any API that has linear complexity.

**Comment by SusanDoggie** [comment]
> > It is easy to separate the interlaced channels when you using raw pointer and stride parameter without copy
> > red: [ red, _ , _, _, red, _ , _, _, red, _ , _, _, ... ]
> > green: [ _, green , _, _, _, green , _, _, _, green , _, _, ... ]
> > blue: [ _, _ , blue, _, _, _ , blue, _, _, _ , blue, _, ...]
> > alpha: [ _, _ , _, alpha, _, _ , _, alpha, _, _ , _, alpha, ... ]
> 
> This particular thing wouldn't stop you from using Collections; you could pretty easily vend a strided view of a base collection like this (just like a slice). But we don't actually _want_ to encourage operating on strided buffers like this in general; it's a performance hazard if you provide any API that has linear complexity.

Yes, I have tried before. that was horribly slow.

**Pull request closed without merging**